rust-src/concordium_base/src/curve_arithmetic/ed25519_instance.rs

```diff
@@ -0,0 +1,245 @@
+    scalar::Scalar,
+    traits::{Identity, VartimeMultiscalarMul, VartimePrecomputedMultiscalarMul},
+};
+
+use super::{Curve, Field, MultiExp, PrimeField};
+
+/// A wrapper to make it possible to implement external traits
+/// and to avoid clashes with blacket implementations.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct RistrettoScalar(Scalar);
+
+impl Serial for RistrettoScalar {
+    fn serial<B: Buffer>(&self, out: &mut B) {
+        let res: &[u8; 32] = self.0.as_bytes();
+        out.write_all(res)
+            .expect("Writing to a buffer should not fail.");
+    }
+}
+
+impl Deserial for RistrettoScalar {
+    fn deserial<R: byteorder::ReadBytesExt>(source: &mut R) -> crate::common::ParseResult<Self> {
+        let mut buf: [u8; 32] = [0; 32];
+        source.read_exact(&mut buf)?;
+        let res = Scalar::from_canonical_bytes(buf).ok_or(anyhow::anyhow!(
+            "Deserialization failed! Not a field value!"
+        ))?;
+        Ok(res.into())
+    }
+}
+
+impl Display for RistrettoScalar {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // Use Debug as Display for now
+        std::fmt::Debug::fmt(self, f)
+    }
+}
+
+// Since we use a wrapper type, it is convenient to use `into()` to convert from
+// Scalar.
+impl From<Scalar> for RistrettoScalar {
+    fn from(value: Scalar) -> Self { RistrettoScalar(value) }
+}
+
+impl Field for RistrettoScalar {
+    fn random<R: rand_core::RngCore + ?std::marker::Sized>(rng: &mut R) -> Self {
+        let mut scalar_bytes = [0u8; 64];
+        rng.fill_bytes(&mut scalar_bytes);
+        Scalar::from_bytes_mod_order_wide(&scalar_bytes).into()
+    }
+
+    fn zero() -> Self { Scalar::zero().into() }
+
+    fn one() -> Self { Scalar::one().into() }
+
+    fn is_zero(&self) -> bool { self.0 == Self::zero().0 }
+
+    fn square(&mut self) { self.0.mul_assign(self.0) }
+
+    fn double(&mut self) { self.0.add_assign(self.0) }
+
+    fn negate(&mut self) {
+        let v = self.0.neg();
+        self.0 = v;
+    }
+
+    fn add_assign(&mut self, other: &Self) { self.0.add_assign(other.0) }
+
+    fn sub_assign(&mut self, other: &Self) { self.0.sub_assign(other.0) }
+
+    fn mul_assign(&mut self, other: &Self) { self.0.mul_assign(other.0) }
+
+    fn inverse(&self) -> Option<Self> {
+        if self.is_zero() {
+            None
+        } else {
+            Some(self.0.invert().into())
+        }
+    }
+}
+
+impl PrimeField for RistrettoScalar {
+    // Taken from `curve25519-dalek` v.4.1.1 that implements `ff::PrimeField`
+    const CAPACITY: u32 = 252;
+    // Taken from `curve25519-dalek` v.4.1.1 that implements `ff::PrimeField``
+    const NUM_BITS: u32 = 253;
+
+    fn into_repr(self) -> Vec<u64> {
+        let mut vec: Vec<u64> = Vec::new();
+        let bytes = self.0.to_bytes();
+        for chunk in bytes.chunks(8) {
+            let x: [u8; 8] = chunk.try_into().unwrap();
```

Comment:
Can you please elaborate here (by documenting) why unwrap is OK.

Reply:
Done.